src/Native/System.Net.Security.Native/pal_gssapi.cpp

```diff
@@ -0,0 +1,265 @@
+    gss_buffer_desc inputNameBuffer{.length = inputNameLen, .value = inputName};
+    gss_OID nameType = const_cast<gss_OID>(GSS_KRB5_NT_PRINCIPAL_NAME);
+    return gss_import_name(minorStatus, &inputNameBuffer, nameType, outputName);
+}
+
+extern "C" uint32_t NetSecurityNative_InitSecContext(uint32_t* minorStatus,
+                                                     GssCredId* claimantCredHandle,
+                                                     GssCtxId** contextHandle,
+                                                     uint32_t isNtlm,
+                                                     GssName* targetName,
+                                                     uint32_t reqFlags,
+                                                     uint8_t* inputBytes,
+                                                     uint32_t inputLength,
+                                                     GssBuffer** outBufferHandle,
+                                                     uint32_t* outTokenLength,
+                                                     uint32_t* retFlags)
+{
+    assert(isNtlm == 0 || isNtlm == 1);
+    assert(minorStatus != nullptr);
+    assert(contextHandle != nullptr);
+    assert(outBufferHandle != nullptr);
+    assert(outTokenLength != nullptr);
+    assert(retFlags != nullptr);
+
+#if HAVE_GSS_SPNEGO_MECHANISM
+    gss_OID desiredMech = isNtlm ? GSS_NTLM_MECHANISM : GSS_SPNEGO_MECHANISM;
+#else
+    assert(!isNtlm && "NTLM is not supported by MIT libgssapi_krb5");
+    gss_OID_desc gss_mech_spnego_OID_desc = {.length = 6, .elements = static_cast<void*>(gss_mech_value)};
+    gss_OID desiredMech = &gss_mech_spnego_OID_desc;
+#endif
+
+    gss_buffer_desc inputToken{.length = UnsignedCast(inputLength), .value = inputBytes};
+    *outBufferHandle = new GssBuffer();
+
+    uint32_t majorStatus = gss_init_sec_context(minorStatus,
+                                                claimantCredHandle,
+                                                contextHandle,
+                                                targetName,
+                                                desiredMech,
+                                                reqFlags,
+                                                0,
+                                                GSS_C_NO_CHANNEL_BINDINGS,
+                                                &inputToken,
+                                                nullptr,
+                                                *outBufferHandle,
+                                                retFlags,
+                                                nullptr);
+
+    NetSecurityNative_HandleError(majorStatus, outBufferHandle, outTokenLength);
+    return majorStatus;
+}
+
+extern "C" uint32_t NetSecurityNative_ReleaseCred(uint32_t* minorStatus, GssCredId** credHandle)
+{
+    assert(minorStatus != nullptr);
+    assert(credHandle != nullptr);
+    return gss_release_cred(minorStatus, credHandle);
```

Comment:
Is this right?  The man page I see says that the second parameter is a ```gss_cred_id_t *cred_handle```.  Is ```gss_cred_id_t``` itself a pointer?  And if it is taking a pointer to a pointer, why / what kind of manipulation is it doing?  I ask because in the one managed call site that calls through to this, it's passing a SafeHandle's handle field by ref and then setting that field to IntPtr.Zero.